File: 2326. Spiral Matrix IV/main.cpp
#include <vector>

using std::pair;
using std::tie;
using std::vector;

struct ListNode {
    int val;
    ListNode *next;
    ListNode()
        : val(0)
        , next(nullptr) {}
    ListNode(int x)
        : val(x)
        , next(nullptr) {}
    ListNode(int x, ListNode *next)
        : val(x)
        , next(next) {}
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode *head) {
        vector<vector<int>> result(m, vector<int>(n, 0));

        int left = 0;
        int right = n - 1;
        int top = 0;
        int bottom = m - 1;

        while (left <= right && top <= bottom) {
            for (int i = left; i <= right; i++) {
                if (head != nullptr) {
                    result[top][i] = head->val;
                    head = head->next;
                } else {
                    result[top][i] = -1;
                }
            }
            top++;

            for (int i = top; i <= bottom; i++) {
                if (head != nullptr) {
                    result[i][right] = head->val;
                    head = head->next;
                } else {
                    result[i][right] = -1;
                }
            }
            right--;

            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    if (head != nullptr) {
                        result[bottom][i] = head->val;
                        head = head->next;
                    } else {
                        result[bottom][i] = -1;
                    }
                }
                bottom--;
            }

            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    if (head != nullptr) {
                        result[i][left] = head->val;
                        head = head->next;
                    } else {
                        result[i][left] = -1;
                    }
                }
                left++;
            }
        }

        return result;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode *head) {
        vector<vector<int>> result(m, vector<int>(n, 0));

        int left = 0;
        int right = n - 1;
        int top = 0;
        int bottom = m - 1;
        int total = m * n;
        int count = 0;
        int direction = 1;

        while (count < total) {
            switch (direction) {
                case 1: {
                    for (int i = left; i <= right; i++) {
                        if (head != NULL) {
                            result[top][i] = head->val;
                            head = head->next;
                        } else {
                            result[top][i] = -1;
                        }
                        count++;
                    }
                    top++;
                    direction = 2;
                    break;
                }
                case 2: {
                    for (int i = top; i <= bottom; i++) {
                        if (head != NULL) {
                            result[i][right] = head->val;
                            head = head->next;
                        } else {
                            result[i][right] = -1;
                        }
                        count++;
                    }
                    right--;
                    direction = 3;
                    break;
                }
                case 3: {
                    for (int i = right; i >= left; i--) {
                        if (head != NULL) {
                            result[bottom][i] = head->val;
                            head = head->next;
                        } else {
                            result[bottom][i] = -1;
                        }
                        count++;
                    }
                    bottom--;
                    direction = 4;
                    break;
                }
                default: {
                    for (int i = bottom; i >= top; i--) {
                        if (head != NULL) {
                            result[i][left] = head->val;
                            head = head->next;
                        } else {
                            result[i][left] = -1;
                        }
                        count++;
                    }
                    left++;
                    direction = 1;
                    break;
                }
            }
        }

        return result;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode *head) {
        vector<vector<int>> result(m, vector<int>(n, -1));

        auto nextCell = [](int row, int col, int direction) -> pair<int, int> {
            switch (direction) {
                case 0:
                    col++;
                    break;
                case 1:
                    row++;
                    break;
                case 2:
                    col--;
                    break;
                default:
                    row--;
                    break;
            }

            return {row, col};
        };

        int row = 0;
        int col = 0;
        int direction = 0;

        while (head != nullptr) {
            result[row][col] = head->val;

            auto [nextRow, nextCol] = nextCell(row, col, direction);
            if (nextRow >= m || nextRow < 0 || nextCol >= n || nextCol < 0 || result[nextRow][nextCol] != -1) {
                direction = (direction + 1) % 4;
            }

            tie(row, col) = nextCell(row, col, direction);
            head = head->next;
        };

        return result;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode *head) {
        vector<vector<int>> result(m, vector<int>(n, -1));
        int row = 0;
        int col = 0;
        int direction = 0;
        vector<vector<int>> dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

        while (head != nullptr) {
            result[row][col] = head->val;

            int nextRow = row + dirs[direction][0];
            int nextCol = col + dirs[direction][1];

            if (nextRow >= m || nextRow < 0 || nextCol >= n || nextCol < 0 || result[nextRow][nextCol] != -1) {
                direction = (direction + 1) % 4;
                nextRow = row + dirs[direction][0];
                nextCol = col + dirs[direction][1];
            }

            row = nextRow;
            col = nextCol;
            head = head->next;
        };

        return result;
    }
};

File: 2326. Spiral Matrix IV/main.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn spiral_matrix(m: i32, n: i32, head: Option<Box<ListNode>>) -> Vec<Vec<i32>> {
        let mut row = 0;
        let mut col = 0;
        let mut direction = 0;
        let mut result = vec![vec![-1; n as usize]; m as usize];
        let mut head = head;
        let dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];

        while let Some(node) = head {
            result[row as usize][col as usize] = node.val;

            let mut next_row = row + dirs[direction][0];
            let mut next_col = col + dirs[direction][1];

            if next_row < 0
                || next_row >= m
                || next_col < 0
                || next_col >= n
                || result[next_row as usize][next_col as usize] != -1
            {
                direction = (direction + 1) % 4;
                next_row = row + dirs[direction][0];
                next_col = col + dirs[direction][1];
            }

            row = next_row;
            col = next_col;

            head = node.next;
        }

        result
    }
}

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn spiral_matrix(m: i32, n: i32, head: Option<Box<ListNode>>) -> Vec<Vec<i32>> {
        let mut left = 0;
        let mut right = n - 1;
        let mut top = 0;
        let mut bottom = m - 1;
        let mut head = head;
        let mut result = vec![vec![-1; n as usize]; m as usize];

        while left <= right && top <= bottom {
            for i in left..=right {
                if let Some(node) = head {
                    result[top as usize][i as usize] = node.val;
                    head = node.next;
                }
            }
            top += 1;

            for i in top..=bottom {
                if let Some(node) = head {
                    result[i as usize][right as usize] = node.val;
                    head = node.next;
                }
            }
            right -= 1;

            if top <= bottom {
                for i in (left..=right).rev() {
                    if let Some(node) = head {
                        result[bottom as usize][i as usize] = node.val;
                        head = node.next;
                    }
                }
                bottom -= 1;
            }

            if left <= right {
                for i in (top..=bottom).rev() {
                    if let Some(node) = head {
                        result[i as usize][left as usize] = node.val;
                        head = node.next;
                    }
                }
                left += 1;
            }
        }

        result
    }
}

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn spiral_matrix(m: i32, n: i32, head: Option<Box<ListNode>>) -> Vec<Vec<i32>> {
        let mut left = 0;
        let mut right = n - 1;
        let mut top = 0;
        let mut bottom = m - 1;
        let mut count = 0;
        let mut head = head;
        let mut direction = 1;
        let mut result = vec![vec![-1; n as usize]; m as usize];
        let total = m * n;

        while count < total {
            match direction {
                1 => {
                    for i in left..=right {
                        if let Some(node) = head {
                            result[top as usize][i as usize] = node.val;
                            head = node.next;
                            count += 1;
                        } else {
                            return result;
                        }
                    }
                    top += 1;
                    direction = 2;
                }
                2 => {
                    for i in top..=bottom {
                        if let Some(node) = head {
                            result[i as usize][right as usize] = node.val;
                            head = node.next;
                            count += 1;
                        } else {
                            return result;
                        }
                    }
                    right -= 1;
                    direction = 3;
                }
                3 => {
                    for i in (left..=right).rev() {
                        if let Some(node) = head {
                            result[bottom as usize][i as usize] = node.val;
                            head = node.next;
                            count += 1;
                        } else {
                            return result;
                        }
                    }
                    bottom -= 1;
                    direction = 4;
                }
                _ => {
                    for i in (top..=bottom).rev() {
                        if let Some(node) = head {
                            result[i as usize][left as usize] = node.val;
                            head = node.next;
                            count += 1;
                        } else {
                            return result;
                        }
                    }
                    left += 1;
                    direction = 1;
                }
            }
        }

        result
    }
}
